scoala-online/main
# Feat/#18 implement subject endpoint

### This pull request solves issue:

Closes: #18 

### Which app is this issue related to?


### Types of changes


### Wiki Updates


### Testing checklist


#### Compatibility tests - **_For E-Learning React App Only_**


### What changes have been done within this issue?

I added the Subject endpoint containing the following components :

- `Subject` located in `model` package
- `SubjectRepository` located in `repository` package
- `SubjectService` located in `service` package
- `SubjectController` located in `controller` package
- `SubjectNotFound` and `SubjectInvalidDataException` located in exception package.

### How should this be manually tested?

You have 2 options:

1. Download and install the neo4j desktop application at 
`https://neo4j.com/download/`

After that. you'll need to create a new local DBMS. You can do that by clicking the `Add` button near `Example Project`
You can choose the name and password but **make sure to remember them**. These will be your username and password

You can start the connection to the new DBMS. After doing that, press `Create database` to make a new, blank database.

2. Create a docker container for your neo4j database as follows
`docker run -p7474:7474 -p7687:7687 neo4j`
You will enter the http://localhost:7474/browser/ and authenticate using:

```
username: neo4j
password: neo4j
```
and reset your password.
> *Note* : To reconnect to the database use
```
$ sudo docker ps -a 
CONTAINER ID   IMAGE         COMMAND                  CREATED        STATUS                    PORTS     NAMES
<some_id>   neo4j:4.2.4   "/sbin/tini -g -- /d…"   21 hours ago   Exited (0) 20 hours ago             local-database1
<some_id>   neo4j:4.2.4   "/sbin/tini -g -- /d…"   27 hours ago   Exited (0) 21 hours ago             local-database2

$ ## find your database
$ docker start local-database1
$ docker attach local-database1
```

After either following method `1` or `2` you need to set up the environment variables of your username and password in the .env file. Please refer to the [Environments wiki page](https://github.com/scoala-online/main/wiki/Environments) if you're unsure how to do it.
Run the java-api application, and use postman to send or receive data.


### Screenshots or example usage
